### PR TITLE
Try out struct based generator encapsulating Rng and Config.

### PR DESCRIPTION
**Pull request by absoludity**
Hi @Kixiron . This is what I think you were describing in https://github.com/vmware/database-stream-processor/pull/101#discussion_r914232253 ?

It adds a `NexmarkGenerator` struct which holds both the `Rng` and the `Config`, so that all the functions no longer need those in their signatures as they're bound to the instance.

I'm still not convinced that it's worth moving from the API defined in the Java Nexmark implementation in this way (and move away from purely functional to maintaining state), but if you think it's worthwhile / important, happy to go with that.

Note that it means we have to use `&mut self` for many of the functions (since the `Rng` needs to be passed as `&mut`).

Anyway, let me know what you think/prefer.

### CODE REVIEW

File: benches/nexmark/generator/mod.rs
Comment (Kixiron) [inline benches/nexmark/generator/mod.rs]: What's the reason for the `?Sized` bound here are any of the rngs we use unsized?

File: benches/nexmark/generator/people.rs
Comment (Kixiron) [inline benches/nexmark/generator/people.rs]: I'd import `Duration` into the file

File: benches/nexmark/generator/people.rs
Comment (Kixiron) [inline benches/nexmark/generator/people.rs]: I'd import `min` into the file

File: benches/nexmark/generator/people.rs
Comment (Kixiron) [inline benches/nexmark/generator/people.rs]: Why do these tests need to be run in serial, it doesn't look like they rely on any global state

**Review by Kixiron — CHANGES_REQUESTED** [review]
Looks pretty good, should be a lot nicer to work with!

File: benches/nexmark/generator/mod.rs
Comment (absoludity) [reply benches/nexmark/generator/mod.rs]: Just from https://docs.rs/rand/latest/rand/trait.Rng.html#generic-usage : 

> The `+ ?Sized` un-bounding allows functions to be called directly on type-erased references; i.e. `foo(r)` where `r: &mut dyn RngCore`. Without this it would be necessary to write `foo(&mut r)`.

Maybe you mean it's not necessary now for the generic struct field? Not sure. I can remove it if you prefer.

File: benches/nexmark/generator/people.rs
Comment (absoludity) [reply benches/nexmark/generator/people.rs]: Answered and fixed on the target PR.

**Comment by absoludity** [comment]
> Looks pretty good, should be a lot nicer to work with!

OK, assuming that you saw my comments in the description, I think you mean that you prefer this structure even if it diverges from the Java implementation's purely functional structure, and will merge this into my other PR once approved.

Thanks again for the reviews.

File: benches/nexmark/generator/mod.rs
Comment (Kixiron) [reply benches/nexmark/generator/mod.rs]: Yah, it's just an unneeded bound in this situation

Comment (absoludity) [reply benches/nexmark/generator/mod.rs]: Ah - k. Removed and rebased on the target again after changes there. Txs.

**Comment by absoludity** [comment]
I you approve this and the target, let's just ensure this merges into the target branch before we then (after it passes checks) merge that upstream. Or otherwise, I can re-create this PR targeting the upstream main and get another +1 review to land. Fine either way. Thanks.

**Review by Kixiron — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
